runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/TransformUpgrader.java

```diff
@@ -0,0 +1,382 @@
+    Map<String, String> inputReplacements = new HashMap<>();
+    if (transformToUpgrade.getOutputsMap().size() == 1) {
+      inputReplacements.put(
+          oldOutputs.iterator().next(),
+          expandedTransform.getOutputsMap().values().iterator().next());
+    } else {
+      for (Map.Entry<String, String> entry : transformToUpgrade.getOutputsMap().entrySet()) {
+        if (expandedTransform.getOutputsMap().keySet().contains(entry.getKey())) {
+          throw new Exception(
+              "Original transform did not have an output with tag "
+                  + entry.getKey()
+                  + " but upgraded transform did.");
+        }
+        String newOutput = expandedTransform.getOutputsMap().get(entry.getKey());
+        if (newOutput == null) {
+          throw new Exception(
+              "Could not find an output with tag "
+                  + entry.getKey()
+                  + " for the transform "
+                  + expandedTransform);
+        }
+        inputReplacements.put(entry.getValue(), newOutput);
+      }
+    }
+
+    String newTransformId = transformId + "_upgraded";
+
+    // The list of obsolete (overridden) transforms that should be removed from the pipeline
+    // produced by this method.
+    List<String> transformsToRemove = new ArrayList<>();
+    recursivelyFindSubTransforms(
+        transformId, runnerAPIpipeline.getComponents(), transformsToRemove);
+
+    Map<String, PTransform> updatedExpandedTransformMap =
+        expandedComponents.getTransformsMap().entrySet().stream()
+            .filter(
+                entry -> {
+                  // Do not include already overridden transforms.
+                  return !transformsToRemove.contains(entry.getKey());
+                })
+            .collect(
+                Collectors.toMap(
+                    entry -> entry.getKey(),
+                    entry -> {
+                      // Fix inputs
+                      Map<String, String> inputsMap = entry.getValue().getInputsMap();
+                      PTransform.Builder transformBuilder = entry.getValue().toBuilder();
+                      if (!Collections.disjoint(inputsMap.values(), inputReplacements.keySet())) {
+                        Map<String, String> updatedInputsMap = new HashMap<>();
+                        for (Map.Entry<String, String> inputEntry : inputsMap.entrySet()) {
+                          String updaterValue =
+                              inputReplacements.containsKey(inputEntry.getValue())
+                                  ? inputReplacements.get(inputEntry.getValue())
+                                  : inputEntry.getValue();
+                          updatedInputsMap.put(inputEntry.getKey(), updaterValue);
+                        }
+                        transformBuilder.clearInputs();
+                        transformBuilder.putAllInputs(updatedInputsMap);
+                      }
+
+                      // Fix sub-transforms
```

Comment:
If you keep the name, you won't have to do this.

Reply:
Done.